scripts/runtests.py

```diff
@@ -0,0 +1,149 @@
+    # strip trailing commas from the file
+    valid_json = re.sub(r",(\s*?[\}\]])", r"\1", "\n".join(lines))
+    return valid_json
+
+
+def _get_strict_params(stub_path: str) -> list[str]:
+    with open(_STRICTER_CONFIG_FILE) as file:
+        data = json.loads(_parse_jsonc(file.read()))
+    if stub_path in data["exclude"]:
+        return []
+    return ["-p", _STRICTER_CONFIG_FILE]
+
+
+if __name__ == "__main__":
+    try:
+        path = sys.argv[1]
+    except IndexError:
+        print("Missing path argument in format <folder>/<stub>", file=sys.stderr)
+        sys.exit(1)
+    path_tokens = Path(path).parts
+    assert len(path_tokens) == 2, "Path argument should be in format <folder>/<stub>"
+    folder, stub = path_tokens
+    assert folder in {"stdlib", "stubs"}, "Only the 'stdlib' and 'stubs' folders are supported"
+    stubtest_result: subprocess.CompletedProcess[bytes] | None = None
+    pytype_result: subprocess.CompletedProcess[bytes] | None = None
+
+    # Run formatters first. Order matters.
+    print("\nRunning pycln...")
+    subprocess.run([sys.executable, "-m", "pycln", path, "--all"])
+    print("\nRunning isort...")
+    subprocess.run([sys.executable, "-m", "isort", path])
+    print("\nRunning Black...")
+    black_result = subprocess.run([sys.executable, "-m", "black", path])
+    if black_result.returncode == 123:
+        print("Could not run tests due to an internal error with Black. See above for details.", file=sys.stderr)
+        sys.exit(black_result.returncode)
+
+    print("\nRunning Flake8...")
+    flake8_result = subprocess.run([sys.executable, "-m", "flake8", path])
+
+    print("\nRunning check_consistent.py...")
+    check_consistent_result = subprocess.run([sys.executable, "tests/check_consistent.py"])
+    print("\nRunning check_new_syntax.py...")
+    check_new_syntax_result = subprocess.run([sys.executable, "tests/check_new_syntax.py"])
+
+    print("\nRunning Pyright...")
+    pyright_result = subprocess.run(
+        [sys.executable, "tests/pyright_test.py", path] + _get_strict_params(path), stderr=subprocess.PIPE, text=True
+    )
+    print(pyright_result.stderr)
+    if re.match(r"error (runn|find)ing npx", pyright_result.stderr):
+        pyright_returncode = 0
+        pyright_skipped = True
+    else:
+        pyright_returncode = pyright_result.returncode
+        pyright_skipped = False
+
+    print("\nRunning mypy...")
+    mypy_result = subprocess.run([sys.executable, "tests/mypy_test.py", path])
+    # If mypy failed, stubtest will fail without any helpful error
+    if mypy_result.returncode == 0:
+        print("\nRunning stubtest...")
+        if folder == "stdlib":
+            stubtest_result = subprocess.run([sys.executable, "tests/stubtest_stdlib.py", stub])
+        else:
+            stubtest_result = subprocess.run([sys.executable, "tests/stubtest_third_party.py", stub])
+    else:
+        print("Skipping stubtest since mypy failed.")
```

Comment:
I wonder if this is possibly a little dangerous. `stubtest_third_party.py` installs packages into an isolated virtual environment, but it's still executing arbitrary code that's been downloaded from PyPI. That's not really an issue when you're doing it in CI (we can set appropriate permissions that mean that there's only so much damage it can do). But maybe it shouldn't be run by default in this script, given that it's the recommended "easy way" for new contributors to run all checks locally, if that's something they want to do.

@hauntsaninja, what do you think?

Reply:
The same issue doesn't really apply to `stubtest_stdlib.py` — we can be reasonably confident that stubtest can import and inspect the stdlib without there being any adverse side effects on the computer stubtest is running on.

Reply:
Btw I'm having issues running stubtest in a virtual environment (".venv"), so I'm already having to run it separately already. I'll raise an issue for that.
Sidenote, there's no mention of such a concern in `tests/README.md`, so if it's valid, it should probably be added there as well.

Reply:
> there's no mention of such a concern in `tests/README.md`, so if it's valid, it should probably be added there as well.

Agreed!

Reply:
I found the issue with running stubtest on windows, it's a venv bug: https://github.com/python/cpython/issues/87915
Basically you HAVE to run using `python` as the command, because:
- `py.exe` does not exists, so that always calls the global installation.
- `python3.exe` is not copied properly in `venv.create`, resulting in the above bug.

Reply:
> I found the issue with running stubtest on windows, it's a venv bug: [python/cpython#87915](https://github.com/python/cpython/issues/87915)

Ew, what a pain.

Reply:
Something like this patch might be a good way of getting round security concerns:

```diff
diff --git a/scripts/runtests.py b/scripts/runtests.py
index 0ee847b9..96b7f73c 100644
--- a/scripts/runtests.py
+++ b/scripts/runtests.py
@@ -91,11 +91,23 @@ def main() -> None:
     mypy_result = subprocess.run([sys.executable, "tests/mypy_test.py", path, "--python-version", _PYTHON_VERSION])
     # If mypy failed, stubtest will fail without any helpful error
     if mypy_result.returncode == 0:
-        print("\nRunning stubtest...")
         if folder == "stdlib":
+            print("\nRunning stubtest...")
             stubtest_result = subprocess.run([sys.executable, "tests/stubtest_stdlib.py", stub])
         else:
-            stubtest_result = subprocess.run([sys.executable, "tests/stubtest_third_party.py", stub])
+            run_stubtest_query = (
+                f"Run stubtest for {stub!r} (Y/N)?\n\n"
+                "NOTE: Running third-party stubtest involves downloading and executing arbitrary code from PyPI.\n"
+                f"Only run stubtest if you trust the {stub!r} package.\n"
+            )
+            run_stubtest_answer = input(colored(run_stubtest_query, "yellow")).lower()
+            while run_stubtest_answer not in {"yes", "no", "y", "n"}:
+                run_stubtest_answer = input(colored("Invalid response; please try again.\n", "red")).lower()
+            if run_stubtest_answer in {"yes", "y"}:
+                print("\nRunning stubtest...")
+                stubtest_result = subprocess.run([sys.executable, "tests/stubtest_third_party.py", stub])
+            else:
+                print(f"Skipping stubtest for {stub!r}...")
```

But maybe having user input halfway through the script would be annoying for your use case in VSCode?

Reply:
In VSCode I'm just loading up a powershell command prompt, not any different than anyone else running the script through their favorite terminal really.

It's more that having the prompt everytime might get annoying and tedious. Could we have a way to auto-accept or auto-deny? Like a command-line argument, or an Env variable? Where you'd have to have read the documentation to even know it's there. Assuring the user still at least (probably) read and accepted the risk.
Should the base `stubtest_third_party` script contain this warning, or is it really not as much of a concern?

Reply:
> Could we have a way to auto-accept or auto-deny? Like a command-line argument, or an Env variable?

That would also be fine!

> Should the base stubtest_third_party script contain this warning, or is it really not as much of a concern?

We should add a note to `tests/README.md` about this, so that contributors know about it.

Reply:
I've filed #8822 to add a warning about the security aspects of running `stubtest_third_party.py` locally, and also improve the docs on a few other things regarding running stubtest locally.